merkle.go

```diff
@@ -0,0 +1,176 @@
+package fastmerkle
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+)
+
+var (
+	errEmptyDataSet = errors.New("empty data set provided")
+)
+
+// GenerateMerkleTree generates a Merkle tree based on the input data
+func GenerateMerkleTree(inputData [][]byte) (*MerkleTree, error) {
+	// Check if the data set is valid
+	if len(inputData) < 1 {
+		return nil, errEmptyDataSet
+	}
+
+	// Create the worker pool and put them on standby
+	workerPool := newWorkerPool(runtime.NumCPU())
+	workerPool.startWorkerPool()
+
+	defer workerPool.close()
+
+	// Generate the leaves of the Merkle tree
+	nodes, leafErr := generateLeaves(inputData, workerPool)
+	if leafErr != nil {
+		return nil, fmt.Errorf(
+			"unable to generate leaf nodes, %w",
+			leafErr,
+		)
+	}
+
+	// While the root is not derived, create new hashing jobs
+	// for the worker pool
+	for len(nodes) > 1 {
+		// Make sure the input node array for the level is even
+		tryDuplicate(&nodes)
```

Comment:
Instead of passing a pointer and internally modifying the input, how about this approach ?

`tryDuplicate -> adjustEvenSize`

```
func adjustEvenSize(nodes []*Node) []*Node {
	if len(nodes)%2 == 0 {
		return nodes
	}

	lastNode := nodes[len(nodes)-1]
	return append(nodes, lastNode.duplicate())
}
```

and in the main loop on this exact line:

`nodes = adjustEvenSize(nodes)` 

cc @Kourin1996

Reply:
I was skeptical at first, but I've benched it and ran the tests - there is a 0 performance difference,  it just makes the code cleaner.

Added 👍 

[8022f4d](https://github.com/Trapesys/fastmerkle/pull/1/commits/8022f4de0e49d515c8aecb6f73e4027980384137)